agent/xds/listeners.go

```diff
@@ -107,17 +107,31 @@ func (s *ResourceGenerator) listenersFromSnapshotConnectProxy(cfgSnap *proxycfg.
 			continue
 		}
 
+		// RDS, Envoy's Route Discovery Service, is only used for HTTP services with a customized discovery chain.
+		useRDS := chain.Protocol != "tcp" && !chain.IsDefault()
+
+		var clusterName string
+		if !useRDS {
+			// When not using RDS we must generate a cluster name to attach to the filter chain.
+			// With RDS, cluster names get attached to the dynamic routes instead.
+			target, err := simpleChainTarget(chain)
```

Comment:
In this block either the protocol is TCP or the chain is default. If the protocol is TCP and the chain is not default then there could be a redirect, but that should be OK since there should only be one node in the chain.

Say you have a service web with a redirect to web in dc2. There would be a single chain node with target `web.default.default.dc2`, which is what we're looking for. The cluster name needs to be for the redirect target because the xds cluster will also have that name: https://github.com/hashicorp/consul/blob/7d95d90fce00199dfdccbaa75eaaaba22c7be698/agent/xds/clusters.go#L610-L612

If there is a resolver with failovers there would also only be a single discovery chain node, despite having multiple chain targets.